lib/services/access_request_test.go

```diff
@@ -19,12 +19,592 @@ package services
 import (
 	"testing"
 
+	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/fixtures"
 	"github.com/gravitational/teleport/lib/utils"
 
+	"github.com/gravitational/trace"
+
+	"github.com/pborman/uuid"
 	"github.com/stretchr/testify/require"
 )
 
+// mockUserAndRoleGetter mocks the UserAndRoleGetter interface.
+type mockUserAndRoleGetter struct {
+	users map[string]types.User
+	roles map[string]types.Role
+}
+
+// user inserts a new user with the specified roles and returns the username.
+func (m *mockUserAndRoleGetter) user(t *testing.T, roles ...string) string {
+	name := uuid.New()
+	user, err := types.NewUser(name)
+	require.NoError(t, err)
+
+	user.SetRoles(roles)
+	m.users[name] = user
+	return name
+}
+
+func (m *mockUserAndRoleGetter) GetUser(name string, withSecrets bool) (types.User, error) {
+	if withSecrets {
+		return nil, trace.NotImplemented("mock getter does not store secrets")
+	}
+	user, ok := m.users[name]
+	if !ok {
+		return nil, trace.NotFound("no such user: %q", name)
+	}
+	return user, nil
+}
+
+func (m *mockUserAndRoleGetter) GetRole(name string) (types.Role, error) {
+	role, ok := m.roles[name]
+	if !ok {
+		return nil, trace.NotFound("no such role: %q", name)
+	}
+	return role, nil
+}
+
+func (m *mockUserAndRoleGetter) GetRoles() ([]types.Role, error) {
+	roles := make([]types.Role, 0, len(m.roles))
+	for _, r := range m.roles {
+		roles = append(roles, r)
+	}
+	return roles, nil
+}
+
+// TestReviewThresholds tests various review threshold scenarios
+func TestReviewThresholds(t *testing.T) {
+	// describes a collection of roles with various approval/review
+	// permissions.
+	roleDesc := map[string]types.RoleConditions{
+		// dictator is the role that will be requested
+		"dictator": {
+			// ...
+		},
+		// populists can achieve dictatorship via uprising or consensus, but only
+		// be conclusively denied via uprising.
+		"populist": {
+			Request: &types.AccessRequestConditions{
+				Roles: []string{"dictator"},
+				Annotations: map[string][]string{
+					"mechanism": {"uprising", "consensus"},
+				},
+				Thresholds: []types.AccessReviewThreshold{
+					{
+						Name:    "overwhelming consensus",
+						Approve: 3,
+					},
```

Comment:
Thresholds without a denial count don't _prevent_ denial.  They just happen to not describe a denial condition (i.e. they don't interfere with other thresholds being capable of triggering denial).  Denial is a privilege which I prefer to have specifically granted. Consider the reverse case:

```
- name: "general disagreement"
  deny: 3
```

It would be *very* bad to silently add an `approve: 1` to that threshold.